### DISCUSSION TITLE
Locales default set at run-time starting with GraalVM/Mandrel 24.2 for JDK 24, how to handle in Quarkus?

### DISCUSSION THREAD

**Discussion by zakkak** [quarkusio/quarkus#43533] [Design Discussions]
With GraalVM/Mandrel prior to 24.2.0 the default locale (and included locale) for a native executable was determined (and implicitly included in the image) by either:

* `user.language` and `user.country`
or
* `quarkus.default-locale`
or
* `quarkus.native.user-language` and `quarkus.native.country` (Deprecated)

settings at native executable build time.

[Starting with GraalVM/Mandrel 24.2.0 for JDK 24](https://github.com/oracle/graal/pull/9694) the build settings of language/country have no effect in terms of included locales (as in the build time locale is not included implicitly) in the resulting native executable. Furthermore, the default locale is determined at runtime based on the locale of the system the native executable runs on, which is on par with how it works in JVM-mode, with the notable difference that if one wants to use a locale not explicitly added to the image it won't work (it will fallback to English).

https://github.com/quarkusio/quarkus/pull/43448 takes a step towards handling this, by automatically including in the native executable the locale defined by either:

* `quarkus.default-locale`
or
* `quarkus.native.user-language` and `quarkus.native.country` (Deprecated)

but doesn't consider `user.language` nor `user.country`.

The questions are:

1. Is that OK?
2. Do we want to add `user.language` and `user.country` properties to the above list?
3. Do we want to include solely the locales explicitly defined through `quarkus.locales` and ignore the build-time default locale?

**Comment by zakkak** [comment]
As an example compiling and running the following code compiled with `native-image -Duser.language=el HelloWorld`:

```java
import java.util.Locale;

public class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello World! The default locale is " + Locale.getDefault());

     for (Locale availableLocale : Locale.getAvailableLocales()) {
        System.out.println(availableLocale);
     }
    }
}
```

gives the following results in the corresponding version:

* Mandrel 23.1
  ```
  ❯ ./helloworld
  Hello World! The default locale is el_IE
  el_IE
  ```
* Mandrel 24.2
  ```
  ❯ ./helloworld                                                
  Hello World! The default locale is en_IE
  
  en_US
  en
  ```

Note that the locale of the system I build on is `en_IE`, that's why setting `user.language` to `el` results in `el_IE`. Also note that despite the build time default locale being `en_IE` the one used at the Mandrel 24.1 generated binary is `en_US` this is because `el_IE` was not included in the image at build time.

If I recompile using `-H:IncludeLocales=en_IE` I am getting:
* Mandrel 23.1
  ```
  ❯ ./helloworld
  Hello World! The default locale is el_IE
  en_ie
  el_IE
  ```
* Mandrel 24.2
  ```
  ❯ ./helloworld                                                
  Hello World! The default locale is en_IE
  
  en_ie
  en_US
  en
  ```

**Reply by turing85** [reply]
Do we - on a technical level - have the possibility to "restore" the old behaviour through some compile-time magic?

**Reply by cescoffier** [reply]
Does it mean it always embeds the default locales during the compilation? 

Could we disable this behavior (at least have a flag for people tuning the set of locales to embed)? My problem is mostly a size increase due to this behavior. 

Now, I believe it makes sense, but as highlighted by @FroMage we need to find the proper way to configure it.

**Reply by zakkak** [reply]
> Do we - on a technical level - have the possibility to "restore" the old behaviour through some compile-time magic?

In theory yes, but I don't think that would be desirable. The new "dynamic" approach is close to the JVM-mode and should be the expected one.

> Does it mean it always embeds the default locales during the compilation?

It seems to always embed the `en_US` locale in addition to anything we configure on top of that.

> Could we disable this behavior (at least have a flag for people tuning the set of locales to embed)? 

As said above, in theory yes, but I am not sure it would be worth or desired.

> My problem is mostly a size increase due to this behavior.

The size impact of an additional locale seems to be on average ~27KB (and that appears to be much lower if you just use latin variations) so it shouldn't be a big deal even for the ones really going for the smaller binary.

**Reply by turing85** [reply]
> > Do we - on a technical level - have the possibility to "restore" the old behaviour through some compile-time magic?
> 
> In theory yes, but I don't think that would be desirable. The new "dynamic" approach is close to the JVM-mode and should be the expected one.

That would have been my next question. And I think that this the core question: do we want some code to provide the "old" behaviour? If so: should it be the default? Or do we just want to have the "new way" and be done with it?

I see where you are coming from. But I think that having the "legacy" behaviour is worth while, since this is a non-trivial change and may lead to nasty bugs for end users.

**Reply by FroMage** [reply]
IMO the build locale should not affect the supported locales of the produced binary. It should be driven entirely by `quarkus.locales` (and transitively by `quarkus.default-locale` which the former defaults to).

**Reply by zakkak** [reply]
So, do you suggest that?

1. if neither `quarkus.locales` nor `quarkus.default-locale` is set, the Quarkus applications should default to English (`en_US` if I am not mistaken) instead of the build systems locale (which is the current behavior) at run-time.

2. if `quarkus.default-locale` is set but `quarkus.locales` is not set, then we should only include the locale `quarkus.default-locale` is set to. This is the current behavior with GraalVM for JDK 21. 

3. if both `quarkus.default-locale` and `quarkus.locales` are set, then we should include only the locales from `quarkus.locales` and the one from `quarkus.default-locale` (this is the current behavior).

4. if `quarkus.locales` is set but `quarkus.default-locale` is not set, then we should include only the locales from `quarkus.locales` and default to English instead of the build systems locale (which is the current behavior) at run-time (similarly to point 1).

For points 2 and 3 starting with graalVM for JDK 24 we will also include `en_US` which shouldn't be a big issue as mentioned in https://github.com/quarkusio/quarkus/discussions/43533#discussioncomment-10795928, but what's interesting is that if the user doesn't specify `user.language` and `user.country` at runtime the application won't start in the `quarkus.default-locale` locale, but instead it will start in the system's default locale (or silently fallback to English if that locale is not embedded in the image). I believe we can change this behavior by checking the value of `quarkus.default-locale` (which is build time fixed) at the app startup and changing the default if desired.

**Reply by FroMage** [reply]
1. yes
2. yes
3. yes
4. yes

Also yes to ignore the current OS locale at startup if `quarkus.default-locale` is set, and use that one instead.

**Reply by jerboaa** [reply]
> I believe we can change this behavior by checking the value of quarkus.default-locale (which is build time fixed) at the app startup and changing the default if desired.

I like this idea.

**Comment by FroMage** [comment]
There's something missing from your explanation, I think: https://quarkus.io/guides/all-config#quarkus-core_quarkus-locales

> [quarkus.locales](https://quarkus.io/guides/all-config#quarkus-core_quarkus-locales)
> The set of supported locales that can be consumed by the extensions.
> The locales must be specified in the IETF BCP 47 format e.g. en-US or fr-FR.
> For instance, the Hibernate Validator extension makes use of it.
> Native-image build uses it to define additional locales that are supposed to be available at runtime.
> A special string "all" is translated as ROOT Locale and then used in native-image to include all locales. Image size penalty applies.

The default value is: `Set containing the build system locale`

Sounds to me like `quarkus.locales` should rather default to `quarkus.default-locale`, not the "build system locale" (which would be what, `user.language` and `user.country`?)

Confusingly, `quarkus.native.user-language` is not described in the all-config page, so I'm not sure what it's supposed to do.

**Reply by zakkak** [reply]
> The default value is: `Set containing the build system locale`
> 
> Sounds to me like `quarkus.locales` should rather default to `quarkus.default-locale`, not the "build system locale" (which would be what, `user.language` and `user.country`?)

Well, that's actually the current Quarkus behavior. See https://github.com/quarkusio/quarkus/blob/96d896763a6e2b7f862bf860b1b36f16f0403c64/core/deployment/src/main/java/io/quarkus/deployment/steps/LocaleProcessor.java#L89-L124 and https://github.com/quarkusio/quarkus/blob/96d896763a6e2b7f862bf860b1b36f16f0403c64/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java#L750-L757

so I guess we should improve the documentation :)


> Confusingly, `quarkus.native.user-language` is not described in the all-config page, so I'm not sure what it's supposed to do.

I guess that's because they are now deprecated. `quarkus.native.user-language` and `quarkus.native.country` were the old way of setting the `quarkus.default-locale` for native images. If set they currently override `quarkus.default-locale` despite being deprecated.

**Reply by FroMage** [reply]
OK, we need to fix the docs, then.

**Comment by brunobat** [comment]
There was a zulip discussion leading to this: https://github.com/quarkusio/quarkus/issues/43578

**Reply by zakkak** [reply]
@brunobat are you suggesting that we should consider a similar approach for locales?

**Reply by brunobat** [reply]
Just that the solution should play ball with this issue.

**Comment by maxandersen** [comment]
Is the rationale for the change in graalvm just "be more like jvm" ? 

what is the actual behavior we think would be nice? 

I'm assuming its one where we can limit how many locales data gets included into the image? Is that still possible in latest graalvm 24?

**Reply by zakkak** [reply]
> Is the rationale for the change in graalvm just "be more like jvm" ?

Yes, but not just for the sake it of it -- it also makes sense to be able to choose any of the included locales as the default one.

> what is the actual behavior we think would be nice?

IMO users should be able to ship their apps with a number of supported locales and have the option to use any of them as the default one.

> I'm assuming its one where we can limit how many locales data gets included into the image? Is that still possible in latest graalvm 24?

Yes, that's still possible. We are just able to also start the application with a different default locale, while with previous versions the default locale was fixed.
